torch/testing/_internal/common_utils.py

```diff
@@ -2448,6 +2448,273 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+
+        If output_tensor is True, yield tensors with the given
+        layout. Otherwise, yield inputs to the corresponding tensor
+        constructors:
+
+          - sparse compressed input is defined as
+            (compressed_indices, plain_indices, values), dict(size=expected_size_from_shape_inference, device=device, dtype=dtype)
+
+          - sparse COO input is defined as
+            (indices, values), dict(size=expected_size_from_shape_inference, device=device, dtype=dtype)
+
+          - strided input is defined as
+            (values,), dict(device=device, dtype=dtype)
+        """
+        if index_dtype is None:
+            index_dtype = torch.int64
+
+        is_compressed_sparse_layout = layout in {torch.sparse_csr, torch.sparse_csc, torch.sparse_bsr, torch.sparse_bsc}
+
+        if output_tensor:
+            for args, kwargs in self.generate_simple_inputs(layout, device=device, dtype=dtype, index_dtype=index_dtype,
+                                                            enable_batched=enable_batched, enable_hybrid=enable_hybrid,
+                                                            enable_zero_sized=enable_zero_sized,
+                                                            enable_batched_variable_nse=enable_batched_variable_nse,
+                                                            output_tensor=False):
+                if layout is torch.strided:
+                    size = kwargs.pop('size', args[0].shape)  # to ensure that a zero-sized tensor has the desired shape
+                    assert len(args) == 1
+                    yield args[0].reshape(size)
+                elif layout is torch.sparse_coo:
+                    yield torch.sparse_coo_tensor(*args, **kwargs)
+                elif is_compressed_sparse_layout:
+                    kwargs.update(layout=layout)
+                    yield torch.sparse_compressed_tensor(*args, **kwargs)
+                else:
+                    assert 0  # unreachable
+            return
+
+        def get_shape(pattern):
+            if isinstance(pattern, list):
+                if not pattern:
+                    return (0,)
+                return (len(pattern), *get_shape(pattern[0]))
+            return ()
+
+        def get_blockpattern(pattern, blocksize):
+            basesize = get_shape(pattern)
+            assert basesize[0] % blocksize[0] == 0, (basesize, blocksize)
+            assert basesize[1] % blocksize[1] == 0, (basesize, blocksize)
+            blockpattern = [[0 for _ in range(basesize[1] // blocksize[1])] for _ in range(basesize[0] // blocksize[0])]
+            for i, row in enumerate(pattern):
+                for j, maskin in enumerate(row):
+                    bi = i // blocksize[0]
+                    bj = j // blocksize[1]
+                    blockno = 1 + bi * (basesize[1] // blocksize[1]) + bj
+                    blockpattern[bi][bj] = blockno if maskin else blockpattern[bi][bj]
```

Comment:
There could be several `i, j` that are mapped to the same `bi, bj`, which means that only the very last "valid" value will be written. It is not a concern here, right? Also nit: seems like this could be simplified with tensors and broadcasting.

Reply:
generate_simple_inputs now uses tensors